From 4a6633deac308a7f3b189746783ff5500f56b67c Mon Sep 17 00:00:00 2001
From: Lothar Rubusch <lothar.rubusch@enclustra.com>
Date: Tue, 20 Sep 2022 20:19:08 +0200
Subject: [PATCH] qspi: fix read instruction

Signed-off-by: Lothar Rubusch <lothar.rubusch@enclustra.com>
---
 drivers/spi/zynq_qspi.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/spi/zynq_qspi.c b/drivers/spi/zynq_qspi.c
index c627c9833e..f4f655b3a6 100644
--- a/drivers/spi/zynq_qspi.c
+++ b/drivers/spi/zynq_qspi.c
@@ -44,7 +44,7 @@ DECLARE_GLOBAL_DATA_PTR;
 #define ZYNQ_QSPI_TXD_00_01_OFFSET	0x80	/* Transmit 1-byte inst */
 #define ZYNQ_QSPI_TXD_00_10_OFFSET	0x84	/* Transmit 2-byte inst */
 #define ZYNQ_QSPI_TXD_00_11_OFFSET	0x88	/* Transmit 3-byte inst */
-#define ZYNQ_QSPI_FR_QOUT_CODE		0x6B	/* read instruction code */
+#define ZYNQ_QSPI_FR_QOUT_CODE		0x6C	/* read instruction code */
 #define ZYNQ_QSPI_FR_DUALIO_CODE	0xBB
 
 /*
-- 
2.25.1

